File: factory/SimpleFactory.java
//factory java

//PizzaStore
abstract class PizzaStore{
	public final Pizza orderPizza(String type){ //order same
		Pizza pizza;
		pizza = createPizza(type);              //
		pizza.prepare();
		pizza.bake();
		pizza.box();
		return pizza;
	}
	abstract Pizza createPizza(String type);    //taste different in other area
}

//New York Pizza Store
class NYPizzaStore extends PizzaStore{
	Pizza createPizza(String item){
		Pizza pizza = null;
		PizzaIngredientFactory ingredientFactory = new NYPizzaIngredientFactory();
		if(item.equals("cheese")){
			pizza = new CheesePizza(ingredientFactory);
			pizza.setName("New York Style Cheese Pizza");
		}
		else if(item.equals("veggie")){
			//...
		}
		return pizza;
	}
}
//Chicago Pizza Store
class ChicagoPizzaStore extends PizzaStore{
	Pizza createPizza(String item){
		Pizza pizza = null;
		PizzaIngredientFactory ingredientFactory = new ChicagoPizzaIngredientFactory();
		if(item.equals("cheese")){
			pizza = new CheesePizza(ingredientFactory);
			pizza.setName("Chicago Style Cheese Pizza");
		}
		else if(item.equals("veggie")){
			//...
		}
		return pizza;
	}
}

//Pizza
abstract class Pizza{
	String name;
	Dough  dough;
	Sauce  sauce;
	Cheese cheese;
	//ArrayList<String> toppings = new ArrayList<String>();
	
	abstract void prepare();   //
	void bake(){
		System.out.println("Bake for 25 minutes...");
	}
	//...
	void box(){
		System.out.println("box done");
	}
	void setName(String name){
		this.name = name;
	}
	public String getName(){
		return name;
	}
	public String toString() {
		StringBuffer result = new StringBuffer();
		result.append("---- " + name + " ----\n");
		if (dough != null) {
			result.append(dough);
			result.append("\n");
		}
		if (sauce != null) {
			result.append(sauce);
			result.append("\n");
		}
		if (cheese != null) {
			result.append(cheese);
			result.append("\n");
		}
		return result.toString();
	}
}

//CheesePizza
//do not need class NYStyleCheesePizza and class ChicagoStyleCheesePizza
class CheesePizza extends Pizza{
	PizzaIngredientFactory ingredientFactory;
	public CheesePizza(PizzaIngredientFactory ingredientFactory){
		this.ingredientFactory = ingredientFactory;
	}
	void prepare(){
		System.out.println("Preparing " + name);
		dough  = ingredientFactory.createDough();
		sauce  = ingredientFactory.createSauce();
		cheese = ingredientFactory.createCheese(); 
	}
}

//ingredient factory
interface PizzaIngredientFactory{
	public Dough  createDough();
	public Sauce  createSauce();
	public Cheese createCheese();
	//...
}

//NYPizzaIngredientFactory
class NYPizzaIngredientFactory implements PizzaIngredientFactory{
	public Dough createDough(){
		return new ThinCrustDough();
	}
	public Sauce createSauce(){
		return new MarinaraSauce();
	}
	public Cheese createCheese(){
		return new ReggianoCheese();
	}
} 

//ChicagoPizzaIngredientFactory
class ChicagoPizzaIngredientFactory implements PizzaIngredientFactory{
	public Dough createDough(){
		return new ThickCrustDough();
	}
	public Sauce createSauce(){
		return new PlumTomatoSauce();
	}
	public Cheese createCheese(){
		return new MozzarellaCheese();
	}
}

//Dough
interface Dough {
	public String toString();
}
class ThinCrustDough implements Dough {
	public String toString() {
		return "Thin Crust Dough";
	}
}
class ThickCrustDough implements Dough {
	public String toString() {
		return "ThickCrust style extra thick crust dough";
	}
}

//Sauce
interface Sauce {
	public String toString();
}
class MarinaraSauce implements Sauce {
	public String toString() {
		return "Marinara Sauce";
	}
}
class PlumTomatoSauce implements Sauce {
	public String toString() {
		return "Tomato sauce with plum tomatoes";
	}
}

//Cheese
interface Cheese {
	public String toString();
}
class MozzarellaCheese implements Cheese {

	public String toString() {
		return "Shredded Mozzarella";
	}
}
class ReggianoCheese implements Cheese {

	public String toString() {
		return "Reggiano Cheese";
	}
}

//SimpleFactory
public class SimpleFactory{
	public static void main(String[] args){
		//1
		PizzaStore nyStore = new NYPizzaStore();
		Pizza      pizza   = nyStore.orderPizza("cheese");
		System.out.println("order a " + pizza);
		
		//2
		PizzaStore chStore = new ChicagoPizzaStore();
		           pizza   = chStore.orderPizza("cheese");
		System.out.println("order a " + pizza);
	}
}

File: factory/SimpleFactory.cpp
//factory c++
#include <iostream>
#include <string>
using namespace std;

//面团 
class Dough{
public:
	virtual string getIngredientInfo() = 0;
};
//薄饼
class ThinCrustDough : public Dough {
public:
	string getIngredientInfo() {
		return "Thin Crust Dough";
	}
};
//厚的
class ThickCrustDough : public Dough {
public:
	string getIngredientInfo(){
		return "ThickCrust style extra thick crust dough";
	}
};

//酱料
class Sauce {
public:
	virtual string getIngredientInfo() = 0;
};
//意大利红酱
class MarinaraSauce : public Sauce {
public:
	string getIngredientInfo(){
		return "Marinara Sauce";
	}
};
//番茄酱
class PlumTomatoSauce : public Sauce {
public:
	string getIngredientInfo() {
		return "Tomato sauce with plum tomatoes";
	}
};

//奶酪
class Cheese {
public:
	virtual string getIngredientInfo() = 0;
};
//莫泽雷勒干酪
class MozzarellaCheese : public Cheese {
public:
	string getIngredientInfo(){
		return "Shredded Mozzarella";
	}
};
//巴马干酪
class ReggianoCheese : public Cheese {
public:
	string getIngredientInfo(){
		return "Reggiano Cheese";
	}
};

//原材料
class PizzaIngredientFactory{
public:
	virtual Dough  * createDough()  = 0;
	virtual Sauce  * createSauce()  = 0;
	virtual Cheese * createCheese() = 0;
	//...
};

//生产 纽约口味
class NYPizzaIngredientFactory : public PizzaIngredientFactory{
public:
	Dough* createDough(){
		return new ThinCrustDough();
	}
	Sauce* createSauce(){
		return new MarinaraSauce();
	}
	Cheese* createCheese(){
		return new ReggianoCheese();
	}
};

//生产 芝加哥口味
class ChicagoPizzaIngredientFactory : public PizzaIngredientFactory{
public: 
	Dough* createDough(){
		return new ThickCrustDough();
	}
	Sauce* createSauce(){
		return new PlumTomatoSauce();
	}
	Cheese* createCheese(){
		return new MozzarellaCheese();
	}
};

//披萨
class Pizza{
public:
	virtual void prepare() = 0;           //统一原材料配给
	
	void bake()                { cout<<"Bake for 25 minutes..."<<endl; }
	void box()                 { cout<<"box done"<<endl; }
	
	void   setName(string name){ this->name = name; }
	string getName()           { return this->name; }
	string getIngredientInfo() {
		string str = "---- ";
			   str += this->name;
			   str += " ----\n";
		if (dough) {
			str += dough->getIngredientInfo() + "\n\r";
		}
		if (sauce) {
			str += sauce->getIngredientInfo() + "\n\r";
		}
		if (cheese) {
			str += cheese->getIngredientInfo() + "\n\r";
		}
		return str;
	}

	string  name;
	Dough * dough;  //饼
	Sauce * sauce;  //酱料
	Cheese* cheese; //奶酪
};

//生产奶酪披萨
class CheesePizza : public Pizza{
public:
	CheesePizza(PizzaIngredientFactory* factory): m_ingredientFactory(factory){}
	void prepare(){
		cout<<"Preparing "<<this->getName()<<endl;
		dough  = m_ingredientFactory->createDough();
		sauce  = m_ingredientFactory->createSauce();
		cheese = m_ingredientFactory->createCheese(); 
	}
private:
	PizzaIngredientFactory* m_ingredientFactory;
};

//披萨店
class PizzaStore{
public:
	virtual Pizza* createPizza(string type) = 0; //不同地域口味的披萨
	Pizza* orderPizza(string type)              //统一品牌流程
	{
		Pizza* pizza = createPizza(type);
		pizza->prepare();
		pizza->bake();
		pizza->box();
		return pizza;
	}
};

//纽约的加盟店
class NYPizzaStore : public PizzaStore{
public:
	Pizza* createPizza(string type){
		Pizza *pizza = NULL;
		PizzaIngredientFactory* ingredientFactory = new NYPizzaIngredientFactory();
		if("cheese" == type){
			pizza = new CheesePizza(ingredientFactory);
			pizza->setName("New York Style Cheese Pizza");
		}
		else if("veggie" == type){
			//...
		}
		return pizza;
	}
};

//芝加哥的
class ChicagoPizzaStore : public PizzaStore{
public:
	Pizza* createPizza(string type){
		Pizza *pizza = NULL;
		PizzaIngredientFactory* ingredientFactory = new ChicagoPizzaIngredientFactory();
		if("cheese" == type){
			pizza = new CheesePizza(ingredientFactory);
			pizza->setName("Chicago Style Cheese Pizza");
		}
		else if("veggie" == type){
			//...
		}
		return pizza;
	}
};

//点餐
int main()
{
	//1
	PizzaStore *nyStore = new NYPizzaStore();
	Pizza      *pizza   = nyStore->orderPizza("cheese");
	cout<<"order a " <<pizza->getIngredientInfo()<<endl;
	//2
	PizzaStore *chStore = new ChicagoPizzaStore();
				pizza   = chStore->orderPizza("cheese");
	cout<<"order a " <<pizza->getIngredientInfo();
	return 0;
}